pgml-extension/src/bindings/transformers/transformers.py

```diff
@@ -956,3 +970,610 @@ def generate(model_id, data, config):
+
+
+class FineTuningBase:
+    def __init__(
+        self,
+        project_id: int,
+        model_id: int,
+        train_dataset: datasets.Dataset,
+        test_dataset: datasets.Dataset,
+        path: str,
+        hyperparameters: dict,
+    ) -> None:
+
+        # initialize class variables
+        self.project_id = project_id
+        self.model_id = model_id
+        self.train_dataset = train_dataset
+        self.test_dataset = test_dataset
+        self.token = None
+        self.load_in_8bit = False
+        self.tokenizer_args = None
+
+        # check if path is a directory
+        if not os.path.isdir(path):
+            os.makedirs(path, exist_ok=True)
+
+        self.path = path
+
+        # check if hyperparameters is a dictionary
+        if "model_name" not in hyperparameters:
+            raise ValueError("model_name is a required hyperparameter")
+        else:
+            self.model_name = hyperparameters.pop("model_name")
+
+        if "token" in hyperparameters:
+            self.token = hyperparameters.pop("token")
+
+        if "training_args" in hyperparameters:
+            self.training_args = hyperparameters.pop("training_args")
+        else:
+            self.training_args = None
+
+        if "project_name" in hyperparameters:
+            project_name = "_".join(hyperparameters.pop("project_name").split())
+            self.training_args["hub_model_id"] = project_name
+
+        if "load_in_8bit" in hyperparameters:
+            self.load_in_8bit = hyperparameters.pop("load_in_8bit")
+
+        if "tokenizer_args" in hyperparameters:
+            self.tokenizer_args = hyperparameters.pop("tokenizer_args")
+
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name, token=self.token
+        )
+
+    def print_number_of_trainable_model_parameters(self, model):
+        """Prints the number of trainable parameters in the model.
+
+        This function traverses all the parameters of a given PyTorch model to
+        count the total number of parameters as well as the number of trainable
+        (i.e., requires gradient) parameters.
+
+        Args:
+            model: A PyTorch model whose parameters you want to count.
+        """
+
+        # Initialize counters for trainable and total parameters
+        trainable_model_params = 0
+        all_model_params = 0
+
+        # Loop through all named parameters in the model
+        for _, param in model.named_parameters():
+            # Update the total number of parameters
+            all_model_params += param.numel()
+
+            # Check if the parameter requires gradient and update the trainable parameter counter
+            if param.requires_grad:
+                trainable_model_params += param.numel()
+
+        # Calculate and print the number and percentage of trainable parameters
+        print_info(f"Trainable model parameters: {trainable_model_params}")
```

Comment:
@kczimm This will require us to use the main thread for ML workloads in our cloud.

Reply:
We need logging visibility during fine tuning.

Reply:
Thanks to a commit by @levkk, we should be able to log from any thread.